one-step-theta.h

```diff
@@ -0,0 +1,110 @@
+#ifndef ONE_STEP_THETA_H
+#define ONE_STEP_THETA_H
+
+#include <deal.II/lac/la_parallel_vector.h>
+#include <deal.II/lac/precondition.h>
+#include <deal.II/lac/solver_gmres.h>
+
+using namespace dealii;
+
+template <typename Operator>
+class OneStepTheta
+{
+public:
+    using Number = double;
+    OneStepTheta(Operator &pde_operator, LinearAlgebra::distributed::Vector<Number> &solution_n);
+
+    template <typename VectorType>
+    void UpdateBuffers(VectorType &solution) const;
+
+    template <typename VectorType>
+    void perform_time_step(double const time_step_size,
+                           VectorType &next_solutions) const;
+
+    void
+    vmult(LinearAlgebra::distributed::Vector<Number> &dst,
+          const LinearAlgebra::distributed::Vector<Number> &src) const;
+
+    template <typename VectorType>
+    void righthandside(VectorType &right_hand_side,
+                       VectorType &current_solution) const;
+
+private:
+    const double theta_ = 0.5;
+    Operator const &pde_operator_;
+    mutable double dt;
+    mutable LinearAlgebra::distributed::Vector<Number> solution_n_;
+};
+
+///// @brief Constructor of the one step theta scheme
+///// @tparam Operator
+///// @param pde_operator
+///// @param solution_n
+template <typename Operator>
+OneStepTheta<Operator>::OneStepTheta(Operator &pde_operator, LinearAlgebra::distributed::Vector<Number> &solution_n) : pde_operator_(pde_operator),
+                                                                                                                       solution_n_(solution_n) {}
+
+///// @brief Left hand operator of the solver
+///// @tparam Operator
+///// @param dst
+///// @param src
+///// @note Operator is phi(u+1)-Theta*(dt*M^-1*R(phi(u+1)))
+template <typename Operator>
+void OneStepTheta<Operator>::vmult(LinearAlgebra::distributed::Vector<Number> &dst,
+                                   const LinearAlgebra::distributed::Vector<Number> &src) const
+{
+    pde_operator_.apply_advection_operator(dst, src);
+    dst *= -1.0;
+    dst *= theta_ * dt;
+    dst.add(1.0, src);
+}
+
+///// @brief Creates the right hand side of the solver
+///// @tparam Operator
+///// @param right_hand_side
+///// @param current_solution
+///// @note Operator is (1-Theta)*(dt*M^-1*R(phi(u)))+u(n)
+template <typename Operator>
+template <typename VectorType>
+void OneStepTheta<Operator>::righthandside(VectorType &right_hand_side,
+                                           VectorType &current_solution) const
```

Comment:
Why do you need to pass `current_solution` again?

Reply:
Because the right hand side of the linear system depends on the current solution and is needed in applying the advection operator
![Screenshot from 2024-02-21 13-52-19](https://github.com/mschreter/2024-MA-Resch/assets/116813975/e01c2b3d-b6d9-417a-9ebd-84d9e0f94aad)

Reply:
I would rename `current_solution` to `old_solution` to make it clear.